File: haskell-server/src/Bolour/Util/FrequencyDistribution.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Bolour.Util.FrequencyDistribution (
    FrequencyDistribution(..)
  , mkFrequencyDistribution
  , mkDistribution
  , randomValue
  , leastFrequentValue
  , normalizedFrequencies
  ) where

import System.Random
import Data.List
import qualified Data.Map as Map

data FrequencyDistribution value = FrequencyDistribution {
    frequencies :: [(value, Int)]
  , distribution :: [(value, Int)]
  , maxDistribution :: Int
  , frequencyMap :: Map.Map value Int
}

mkFrequencyDistribution :: Ord value => [(value, Int)] -> FrequencyDistribution value
mkFrequencyDistribution frequencies =
  let distribution = mkDistribution frequencies
      maxDistribution = snd $ last distribution
      frequencyMap = Map.fromList frequencies
  in FrequencyDistribution frequencies distribution maxDistribution frequencyMap

mkDistribution :: [(value, Int)] -> [(value, Int)]
mkDistribution frequencies =
  let nextDistribution (val, total) (nextVal, frequency) = (nextVal, total + frequency)
  in scanl1 nextDistribution frequencies

randomValue :: FrequencyDistribution value -> IO value
randomValue FrequencyDistribution {distribution, maxDistribution} = do
  dist <- randomRIO (0, maxDistribution - 1) -- uses closed interval [0, maxDistribution - 1]
  -- Invariant dist < maxDistribution => there exists an index whose distribution value is > dist.
  let Just index = findIndex ((<) dist . snd) distribution
      val = fst $ distribution !! index
  return val

-- | Get the least frequent value.
leastFrequentValue :: Ord value => FrequencyDistribution value -> [value] -> Maybe value
leastFrequentValue FrequencyDistribution {frequencyMap} vs =
  let freq v = Map.lookup v frequencyMap
      minFreq v1 v2 = if freq v1 <= freq v2 then v1 else v2
  in case vs of
    [] -> Nothing
    _ -> Just $ foldl1' minFreq vs

normalizedFrequencies :: Ord value => FrequencyDistribution value -> Int -> ((Map.Map value Int), Int)
normalizedFrequencies FrequencyDistribution {maxDistribution, frequencyMap} roughTotal =
  let factor :: Float = fromIntegral roughTotal / fromIntegral maxDistribution
      normalizer freq = max 1 (round $ fromIntegral freq * factor)
      normalized = normalizer <$> frequencyMap
      total = sum $ Map.elems normalized
  in (normalized, total)

File: scala-server/scalautil/src/main/scala/com/bolour/util/FrequencyDistribution.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
 */

package com.bolour.util

/**
  * Frequency distribution for a list of values.
  *
  * @param frequencies The frequencies: represented for each value as a non-negative integer.
  */
case class FrequencyDistribution[T](frequencies: List[(T, Int)]) {

  val frequencyMap = Map(frequencies:_*)

  /**
    * Cumulative distribution function.
    */
  val distribution = frequencies.tail.scanLeft(frequencies.head) (
    (cumulative, element) => (cumulative, element) match {
      case ((_, cumWeight), (nextVal, weight)) =>
        (nextVal, cumWeight + weight)
    }
  )

  val maxDistribution: Int = distribution.last._2

  /**
    * Get a random value according to the frequencies in this distribution.
    */
  def randomValue(): T = {
    val distLevel: Int = (Math.random() * maxDistribution).toInt
    distribution.find(_._2 >= distLevel).get._1
  }

  /**
    * Get the least frequent value among a given set of values.
    */
  def leastFrequentValue(values: List[T]): Option[T] = {
    def freq(value: T) = frequencyMap(value)
    def least(vs: List[T]): T =
      vs.reduce( (v1, v2) => if (freq(v1) <= freq(v2)) v1 else v2 )
    values match {
      case Nil => None
      case _ => Some(least(values))
    }
  }

  /**
    * Normalize the frequencies to obtain a given rough total
    * return the normalized frequencies and the actual total.
    */
  def normalizedFrequencies(roughTotal: Int): (Map[T, Int], Int) = {
    val factor: Float = roughTotal.toFloat / maxDistribution.toFloat
    def normalizer(frequency: Int) = Math.round(frequency * factor).max(1)

    val normalized = frequencyMap mapValues normalizer
    val actualTotal = normalized.values.sum
    (normalized, actualTotal)
  }
}
